apps/src/netsim/NetSimRouterNode.js

```diff
@@ -925,85 +917,76 @@ var contains = function (haystack, needle) {
  *        if connection is allowed, FALSE if connection is rejected.
  */
 NetSimRouterNode.prototype.acceptConnection = function (otherNode, onComplete) {
-  var self = this;
-  this.countConnections(function (err, count) {
-    if (err) {
-      onComplete(err, false);
-      return;
-    }
-
-    if (count > MAX_CLIENT_CONNECTIONS) {
-      onComplete(new Error("Too many connections"), false);
-      return;
-    }
-
-    // Trigger an update, which will correct our connection count
-    self.update(function (err) {
-      onComplete(err, err === null);
-    });
-  });
+  var rejectionReason = null;
+
+  // Force a refresh to verify that we have not exceeded the connection limit.
+  this.shard_.wireTable.refresh()
+      .done(function () {
+        var connections = this.getConnections();
+
+        // Check for connection limit exceeded
+        if (connections.length > this.maxClientConnections_) {
+          rejectionReason = new Error("Too many connections.");
+          return;
+        }
+
+        // Check for address collisions
+        var addressesSoFar = {};
+        addressesSoFar[this.getAddress()] = true;
+        addressesSoFar[this.getAutoDnsAddress()] = true;
+        var addressCollision = connections.some(function (wire) {
+          var collides = addressesSoFar.hasOwnProperty(wire.localAddress);
+          addressesSoFar[wire.localAddress] = true;
+          return collides;
+        });
+        if (addressCollision) {
+          rejectionReason = new Error("Address collision detected.");
+        }
+
+      }.bind(this))
+      .fail(function (err) {
+        rejectionReason = err;
+      })
+      .always(function () {
+        onComplete(rejectionReason, null === rejectionReason);
+      });
 };
 
 /**
- * Assign a new address for hostname on wire, calling onComplete
- * when done.
- * @param {!NetSimWire} wire that lacks addresses or hostnames
- * @param {string} hostname of requesting node
- * @param {NodeStyleCallback} [onComplete]
+ * Generate a list of available addresses, then pick one at random and return it.
+ * @returns {string} a new available address.
  */
-NetSimRouterNode.prototype.requestAddress = function (wire, hostname, onComplete) {
-  onComplete = onComplete || function () {};
-
-  // General strategy: Create a list of existing remote addresses, pick a
-  // new one, and assign it to the provided wire.
-  var self = this;
-  this.getConnections(function (err, wires) {
-    if (err) {
-      onComplete(err);
-      return;
-    }
-
-    var addressList = wires.filter(function (wire) {
-      return wire.localAddress !== undefined;
-    }).map(function (wire) {
-      return wire.localAddress;
-    });
+NetSimRouterNode.prototype.getRandomAvailableClientAddress = function () {
```

Comment:
Explicit feature request from the ed team - in lessons designed to motivate DNS it is too easy to guess hidden user addresses when they are sequential, especially in small groups.

Initially implemented in PR #2540